NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,483 @@
+			sqvm,
+			fmt::format(
+				"Congrats, {}!\nYou have reached the maximum folder size!\n\n... please delete something. Or make your file shorter. "
+				"Either works.",
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	g_saveFileManager->SaveFileAsync<context>(dir / fileName, content);
+
+	return SQRESULT_NULL;
+}
+
+#include "scripts/scriptjson.h"
+
+// void NSSaveJSONFile(string file, table data)
+ADD_SQFUNC("void", NSSaveJSONFile, "string file, table data", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm);
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	fs::path dir = savePath / fs::path(mod->m_ModDirectory).filename();
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (!IsPathSafe(fileName, dir))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format("File name invalid ({})! Make sure it results in a path inside your mod's save folder.", fileName, mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	std::string content = EncodeJSON<context>(sqvm);
+	if (ContainsInvalidChars(content))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm, fmt::format("File contents may not contain NUL/\\0 characters! Make sure your strings are valid!", mod->Name).c_str());
+		return SQRESULT_ERROR;
+	}
+
+	fs::create_directories(dir);
+	// this actually allows mods to go over the limit, but not by much
+	// the limit is to prevent mods from taking gigabytes of space,
+	// this ain't a cloud service.
+	if (GetSizeOfFolderContentsMinusFile(dir, fileName) + content.length() > MAX_FOLDER_SIZE)
+	{
+		// tbh, you're either trying to fill the hard drive or use so much data, you SHOULD be congratulated.
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format(
+				"Congrats, {}!\nYou have reached the maximum folder size!\n\n... please delete something. Or make your file shorter. "
+				"Either works.",
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	g_saveFileManager->SaveFileAsync<context>(dir / fileName, content);
+
+	return SQRESULT_NULL;
+}
+
+// int NS_InternalLoadFile(string file)
+ADD_SQFUNC("int", NS_InternalLoadFile, "string file", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm, 1); // the function that called NSLoadFile :)
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	fs::path dir = savePath / fs::path(mod->m_ModDirectory).filename();
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (!IsPathSafe(fileName, dir))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format("File name invalid ({})! Make sure it results in a path inside your mod's save folder.", fileName, mod->Name)
```

Comment:
Another reason for this to be raised is the usage of non-ASCII characters, I think it should be mentioned here:
```suggestion
			fmt::format("File name invalid ({})! Make sure it does not contain any non-ASCII character, and results in a path inside your mod's save folder.", fileName, mod->Name)
```